pkg/config/release.go

```diff
@@ -163,6 +183,278 @@ func GetChangedTemplates(path, baseRev string) ([]ConfigMapSource, error) {
+				Steps: chain,
+			},
+		}
+		return change, nil
+	}
+	if strings.HasSuffix(filename, "-workflow.yaml") {
+		name := strings.TrimSuffix(filename, "-workflow.yaml")
+		workflow, ok := fullRegistry.workflows[name]
+		if !ok {
+			return RegistryStep{}, fmt.Errorf("could not find workflow: %s", name)
+		}
+		change := RegistryStep{
+			Workflow: &api.RegistryWorkflow{
+				As:    name,
+				Steps: workflow,
+			},
+		}
+		return change, nil
+	}
+	return RegistryStep{}, fmt.Errorf("invalid step filename: %s", filename)
+}
+
+// GetChangedRegistrySteps identifies all registry components (refs, chains, and workflows) that changed.
+// After all changes have been identified, the list gets "trimmed": all changed refs that are included in
+// changed chains are removed from the list since the refs will be tested when the chain gets tested. This
+// is also done for changed chains and refs included in changed workflows. This helps reduce the number of
+// rehearsals that are performed.
+func GetChangedRegistrySteps(path, baseRev string, refs map[string]api.LiteralTestStep, chains map[string][]api.TestStep, workflows map[string]api.MultiStageTestConfiguration) ([]RegistryStep, error) {
+	revChanges, err := getRevChanges(path, RegistryPath, baseRev, true)
+	if err != nil {
+		return nil, err
+	}
+	var registryChanges []RegistryStep
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load registry: %v", err)
+	}
+	fullRegistry := registry{refs: refs, chains: chains, workflows: workflows}
+	seenReference := make(map[string]bool)
+	for _, c := range revChanges {
+		if filepath.Ext(c.Filename) == ".yaml" || strings.HasSuffix(c.Filename, "-commands.sh") {
+			change, err := loadRegistryStep(filepath.Base(c.Filename), fullRegistry)
+			if err != nil {
+				return nil, err
+			}
+			if change.Reference != nil {
+				if _, ok := seenReference[change.Reference.As]; ok {
+					continue
+				}
+				seenReference[change.Reference.As] = true
+			}
+			registryChanges = append(registryChanges, change)
+		}
+	}
+	trimmedChanges := trimRegistryChanges(registryChanges, fullRegistry)
+	return trimmedChanges, nil
+}
+
+// getNestedChainsFromSteps returns the names of all chains nested inside other chains in the provided RegistryChanges
+func getNestedChainsFromSteps(chains []api.TestStep, fullRegistry registry) []string {
+	// make a set of chain names to prevent duplicates
+	nestedChains := sets.NewString()
+	for _, change := range chains {
+		steps, ok := fullRegistry.chains[*change.Chain]
+		if !ok {
+			// handle this
+		}
+		for _, step := range steps {
+			// ignore steps that aren't chains
+			if step.Chain == nil {
+				continue
+			}
+			// add nested chain to set
+			nestedChains.Insert(*step.Chain)
+			// recursively check for nested chains and add those to the nested chain set
+			for _, chain := range getNestedChainsFromSteps(fullRegistry.chains[*step.Chain], fullRegistry) {
+				nestedChains.Insert(chain)
+			}
+		}
+	}
+	// convert set into slice
```

Comment:
you *could* return `nestedChains.List()` but I suggest you return the set since you can use it's methods later